app/src/org/commcare/connect/ConnectIdWorkflows.java

```diff
@@ -435,27 +440,28 @@ public static boolean handleFinishedActivity(int requestCode, int resultCode, In
                 }
             }
             case CONNECT_RECOVERY_ALT_PHONE_MESSAGE -> {
-                if (success) {
-                    nextRequestCode = ConnectTask.CONNECT_RECOVERY_VERIFY_ALT_PHONE;
-                }
-            }
-            case CONNECT_VERIFY_ALT_PHONE_MESSAGE -> {
                 if (success) {
                     boolean change = intent.getBooleanExtra(ConnectConstants.BUTTON2, false);
 
-                    nextRequestCode = change ? ConnectTask.CONNECT_VERIFY_ALT_PHONE_CHANGE : ConnectTask.CONNECT_VERIFY_ALT_PHONE;
+                    nextRequestCode = change ? ConnectTask.CONNECT_RECOVERY_VERIFY_PIN : ConnectTask.CONNECT_RECOVERY_VERIFY_ALT_PHONE;
                 }
             }
-            case CONNECT_RECOVERY_VERIFY_ALT_PHONE -> {
-                nextRequestCode = success ? ConnectTask.CONNECT_RECOVERY_CHANGE_PIN :
-                        ConnectTask.CONNECT_RECOVERY_VERIFY_PRIMARY_PHONE;
+            case CONNECT_RECOVERY_WRONG_PIN -> {
+                if (success) {
+                    nextRequestCode = ConnectManager.getFailureAttempt()>2  ? ConnectTask.CONNECT_RECOVERY_VERIFY_ALT_PHONE : ConnectTask.CONNECT_RECOVERY_VERIFY_PIN;
+                }
             }
-            case CONNECT_VERIFY_ALT_PHONE_CHANGE -> {
-                nextRequestCode = success ? ConnectTask.CONNECT_VERIFY_ALT_PHONE : ConnectTask.CONNECT_VERIFY_ALT_PHONE_MESSAGE;
+            case CONNECT_REGISTRATION_WRONG_PIN -> {
+                if (success) {
+                    nextRequestCode = ConnectManager.getFailureAttempt()>2 ? ConnectTask.CONNECT_REGISTRATION_CHANGE_PIN : ConnectTask.CONNECT_REGISTRATION_CONFIRM_PIN;
+                }
             }
+            case CONNECT_RECOVERY_VERIFY_ALT_PHONE -> nextRequestCode = success ? ConnectTask.CONNECT_RECOVERY_CHANGE_PIN :
+                    ConnectTask.CONNECT_RECOVERY_VERIFY_PRIMARY_PHONE;
+            case CONNECT_VERIFY_ALT_PHONE_CHANGE -> nextRequestCode = success ? ConnectTask.CONNECT_VERIFY_ALT_PHONE : ConnectTask.CONNECT_VERIFY_ALT_PHONE_MESSAGE;
             case CONNECT_VERIFY_ALT_PHONE -> {
                 if(success) {
-                    completeSignin();
+                    completeSignIn();
                 }
             }
             case CONNECT_RECOVERY_CHANGE_PIN -> {
```

Comment:
FIX: A couple lines below here, change CONNECT_RECOVERY_VERIFY_PRIMARY_PHONE to CONNECT_RECOVERY_VERIFY_ALT_PHONE